### PR TITLE
apple, core, server: app store billing

### PR DESCRIPTION
**Pull request by SmailBarkouch**
Apple users can now purchase premium via App Store billing. This PR is split into two parts:
- Front-end UI work on the apple clients. The majority of this UI is found in the `clients/apple/Shared` folder.
- Back-end UI work on core and server. This includes validations when first upgrading to premium, and processing app store server notifications.

### New Server Endpoints
- `/upgrade-account-app-store`: Accepts the original transaction id (first transaction id), app account token (used to link a lockbook account with an App Store account), and an encoded receipt. The encoded receipt is sent to apple to for validation, and the returned details will be checked up against the app account token and original transaction id.
- `/app_store_notification_webhook`: Processes app store server notifications.

### Code Organization
- `server/.../billing/app_store_client.rs`: Builds the requests that are made to apple's servers. This includes creating a JWS for authentication.
- `server/.../billing/app_store_service.rs`: Helper functions for `upgrade_account_app_store` and `app_store_notification_webhook`.
- `server/.../billing/app_store_model.rs`: The data types that are needed for parsing the responses and request bodys' of apple requests.

### Progress
- [x] Create iOS UI
- [ ] Create MacOS UI
- [x] Implement `upgrade_account_app_store` endpoint
- [x] Implement `app_store_notifications_webhook` endpoint
- [x] Test billing for iOS
- [ ] Test billing for MacOS
- [ ] Test edge case billing operations on app store server notifications
- [ ] screenshots in docs

### CODE REVIEW

File: clients/apple/Shared/Stateful Logic/BillingService.swift [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+    var makingPurchaseAttempt = false
+    
+    
+    init(_ core: LockbookApi) {
+        self.core = core
+        updateListenerTask = listenForTransactions()
+
+        Task {
+            await requestProducts()
+        }
```
Comment (Parth) [inline clients/apple/Shared/Stateful Logic/BillingService.swift:42]: I wonder if we don't want to do this every time the app opens

File: clients/apple/Shared/Stateful Logic/BillingService.swift [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+            if storeProducts.count == 1 {
+                maybeMonthlySubscription = storeProducts[0]
+            } else {
+                print("No products!")
+            }
+            
+        } catch {
+            print("Failed product request from the App Store server: \(error)")
+        }
+    }
```
Comment (Parth) [inline clients/apple/Shared/Stateful Logic/BillingService.swift:172]: I wonder if this fails on startup, will something later on act unexpectedly

File: clients/apple/Shared/Stateful Logic/BillingService.swift [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+                }
+            }
+        }
+    }
+
+    
+    func purchasePremium() async throws -> PurchaseResult? {
+        let accountToken = UUID()
+        let purchaseOpt: Set<Product.PurchaseOption> = [Product.PurchaseOption.appAccountToken(accountToken)]
+
```
Comment (Parth) [inline clients/apple/Shared/Stateful Logic/BillingService.swift:89]: could await requestProducts right here right?

Comment (Parth) [reply clients/apple/Shared/Stateful Logic/BillingService.swift:89]: instead of `init`

File: clients/apple/Shared/Stateful Logic/BillingService.swift [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+            guard let monthlySubscription = maybeMonthlySubscription else {
+                throw StoreError.noProduct
+            }
+            let result = try await monthlySubscription.purchase(options: purchaseOpt)
+            SKReceiptRefreshRequest().start()
+            
+            switch result {
+            case .success(let verification):
+                while true {
+                    if let receipt = getReceipt() {
```
Comment (Parth) [inline clients/apple/Shared/Stateful Logic/BillingService.swift:100]: Can there be no receipt? Wouldn't this inf loop if no receipt?

File: clients/apple/Shared/Stateful Logic/BillingService.swift [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+                        makingPurchaseAttempt = true
+                        let transaction = try checkVerified(verification)
```
Comment (Parth) [inline clients/apple/Shared/Stateful Logic/BillingService.swift:102]: what happens if checkVerified throws?

File: clients/apple/Shared/Stateful Logic/BillingService.swift [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+            print("Couldn't read receipt data with error: " + error.localizedDescription)
+        }
+
+        return .failure
+    }
+    
+    func checkVerified<T>(_ result: VerificationResult<T>) throws -> T {
+        switch result {
+        case .unverified:
+            throw StoreError.failedVerification
```
Comment (Parth) [inline clients/apple/Shared/Stateful Logic/BillingService.swift:138]: Why does this need to throw?

File: clients/apple/Shared/Stateful Logic/BillingService.swift [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+                while true {
+                    if let receipt = getReceipt() {
+                        makingPurchaseAttempt = true
+                        let transaction = try checkVerified(verification)
+                        
+                        let result = self.core.upgradeAccountAppStore(originalTransactionId: String(transaction.originalID), appAccountToken: accountToken.uuidString.lowercased(), encodedReceipt: receipt)
+                        
+                        switch result {
+                        case .success(_):
+                            await transaction.finish()
```
Comment (Parth) [inline clients/apple/Shared/Stateful Logic/BillingService.swift:108]: pretty nested at this point, I think there's probably a way to write this all to be more readable

File: clients/apple/Shared/Views/ManageSubscription.swift [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+            }
+            .padding(.vertical)
+                
+            switch settings.tier {
+            case .Trial: trial
+            case .Premium: trial
+            case .Unknown: trial
+            }
+            
+            Text("Expand your storage to **30GB** for just **\(billing.maybeMonthlySubscription!.displayPrice)** a month.")
```
Comment (Parth) [inline clients/apple/Shared/Views/ManageSubscription.swift:27]:
```suggestion
            Text("Expand your storage to **30GB** for **\(billing.maybeMonthlySubscription!.displayPrice)** a month.")
```

What language are we using elsewhere?

File: clients/apple/Shared/Views/ManageSubscription.swift [hunk 9]
```diff
@@ -0,0 +1,118 @@
+import SwiftUI
```
Comment (Parth) [inline clients/apple/Shared/Views/ManageSubscription.swift:1]: Will re-review this after screenshots

File: clients/apple/Shared/Views/ManageSubscription.swift [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+import AlertToast
+
+struct ManageSubscription: View {    
+    @EnvironmentObject var billing: BillingService
+    @EnvironmentObject var settings: SettingsService
+    
+    @Environment(\.presentationMode) var presentationMode
+        
+    @State var isLoading = false
```
Comment (Parth) [inline clients/apple/Shared/Views/ManageSubscription.swift:10]: prob can be more specific here

File: clients/apple/Shared/Views/ManageSubscription.swift [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+    
+    @ViewBuilder
+    var trial: some View {
+        VStack(alignment: .leading) {
+            Text("If you upgraded, your usage would be:")
+            ColorProgressBar(value: settings.premiumProgress)
+        }
+    }
+}
+
```
[the remark this answers is not in the document]
Comment (Parth) [reply clients/apple/Shared/Views/ManageSubscription.swift:110]: Loading, errors, etc

File: containers/local.env [hunk 12]
```diff
@@ -19,6 +19,12 @@ MILLIS_BETWEEN_METRICS=100
 
 ADMINS="admin1"
 
+APPLE_SUBSCRIPTION_PRODUCT_ID="basic.premium"
+APPLE_IAP_KEY=${APPLE_IAP_KEY:=1234}
+APPLE_IAP_KEY_ID=${APPLE_IAP_KEY_ID:=1234}
+APPLE_ISSUER_ID=${APPLE_ISSUER_ID:=1234}
+APP_STORE_CONNECT_SHARED_SECRET=${APP_STORE_CONNECT_SHARED_SECRET:=sharedsecret}
```
Comment (Parth) [inline containers/local.env:26]: make this start with apple too? 

These can prob be shorter

File: core/libs/shared/src/api.rs [hunk 13]
```diff
@@ -389,6 +389,32 @@ impl Request for UpgradeAccountGooglePlayRequest {
     const ROUTE: &'static str = "/upgrade-account-google-play";
 }
 
+#[derive(Serialize, Deserialize, Debug, PartialEq, Eq, Clone)]
+pub struct UpgradeAccountAppStoreRequest {
+    pub original_transaction_id: String,
```
Comment (Parth) [inline core/libs/shared/src/api.rs:394]: what is this "original" business?

File: server/server/src/billing/app_store_client.rs [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+use crate::config::AppleConfig;
+use jsonwebtoken::{encode, Algorithm, EncodingKey, Header};
+use lockbook_shared::clock::get_time;
+use reqwest::RequestBuilder;
+use serde::{Deserialize, Serialize};
+
+const VERIFY_PROD: &str = "https://buy.itunes.apple.com/verifyReceipt";
+const VERIFY_SANDBOX: &str = "https://sandbox.itunes.apple.com/verifyReceipt";
+
+const ALG: &str = "ES256";
```
Comment (Parth) [inline server/server/src/billing/app_store_client.rs:11]: can this change?

File: server/server/src/billing/app_store_client.rs [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+const TYP: &str = "JWT";
+const AUDIENCE: &str = "appstoreconnect-v1";
+const BUNDLE_ID: &str = "app.lockbook";
+
+#[derive(Debug)]
+pub enum AppStoreError {
+    Other(String),
+}
```
Comment (Parth) [inline server/server/src/billing/app_store_client.rs:19]: seems weird

File: server/server/src/billing/app_store_client.rs [hunk 16]
```diff
@@ -0,0 +1,92 @@
...
+            let resp_body: VerifyReceiptResponse = resp.json().await?;
+
+            if resp_body.status == 0 {
+                Ok(resp_body)
+            } else {
+                Err(AppStoreError::Other(format!("Unexpected response: {:?}", resp_body)))
+            }
+        }
+        _ => Err(AppStoreError::Other(format!("Unexpected response: {:?}", resp_body))),
+    }
```
Comment (Parth) [inline server/server/src/billing/app_store_client.rs:91]: should never happen in prod right?

File: server/server/src/schema.rs [hunk 17]
```diff
@@ -45,7 +45,8 @@ pub mod v2 {
             file_children: <Uuid, HashSet<Uuid>>,
             sizes: <Uuid, u64>,
             google_play_ids: <String, Owner>,
-            stripe_ids: <String, Owner>
+            stripe_ids: <String, Owner>,
+            app_store_ids: <String, Owner>
```
Comment (Parth) [inline server/server/src/schema.rs:49]: I think we'll need a migration

File: server/server/src/router_service.rs [hunk 18]
```diff
@@ -268,6 +269,47 @@ pub fn google_play_notification_webhooks(
...
+    server_state: &Arc<ServerState>,
+) -> impl Filter<Extract = impl warp::Reply, Error = warp::Rejection> + Clone {
+    let cloned_state = server_state.clone();
+
+    warp::post()
+        .and(warp::path(APP_STORE_WEBHOOK_ROUTE))
+        .and(warp::any().map(move || cloned_state.clone()))
+        .and(warp::body::bytes())
+        .then(|state: Arc<ServerState>, body: Bytes| async move {
+            let span = span!(
```
Comment (Parth) [inline server/server/src/router_service.rs:283]: Can you make sure all the webhooks are getting timed? Just once over and compare to the main `core_req` and make sure we're doing everything the same way, logging & timing. Make sure we're creating spans the same way and that we're timing the same way.

File: server/server/src/config.rs [hunk 19]
```diff
@@ -213,6 +214,36 @@ impl BillingConfig {
             ),
             google: GoogleConfig::from_env_vars(),
             stripe: StripeConfig::from_env_vars(),
+            apple: AppleConfig::from_env_vars(),
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct AppleConfig {
```
Comment (Parth) [inline server/server/src/config.rs:223]: Feels like we're out growing env vars, maybe we want to start using toml for config in the future.

File: server/server/src/config.rs [hunk 20]
```diff
@@ -213,6 +214,36 @@ impl BillingConfig {
...
+        let apple_root_cert_dest = env_or_empty("APPLE_ROOT_CERT_PATH");
+
+        Self {
+            iap_key: env_or_panic("APPLE_IAP_KEY"),
+            iap_key_id: env_or_panic("APPLE_IAP_KEY_ID"),
+            asc_public_key: env_or_panic("APPLE_APP_STORE_CONNECT_PUBLIC_KEY"),
+            issuer_id: env_or_panic("APPLE_ISSUER_ID"),
+            subscription_product_id: env_or_panic("APPLE_SUBSCRIPTION_PRODUCT_ID"),
+            asc_shared_secret: env_or_panic("APP_STORE_CONNECT_SHARED_SECRET"),
+            apple_root_cert: apple_root_cert_dest
```
Comment (Parth) [inline server/server/src/config.rs:244]: how large is this? Can't just stuff it in an env var right?

Comment (Parth) [reply server/server/src/config.rs:244]: Oh it's non-utf-8 or something right?

Comment (Parth) [reply server/server/src/config.rs:244]: We could base64 encode it or something, want to avoid random files we have to keep track of.

File: server/server/src/billing/billing_service.rs [hunk 21]
```diff
@@ -73,6 +79,60 @@ fn release_subscription_profile<T: Debug>(
     })?)
 }
 
+pub async fn upgrade_account_app_store(
```
Comment (Parth) [inline server/server/src/billing/billing_service.rs:82]: ideal top level function, very readable

File: server/server/src/billing/app_store_service.rs [hunk 22]
```diff
@@ -0,0 +1,131 @@
...
+                        .subject_pki,
+                ))
+                .map_err(|_| ClientError(AppStoreNotificationError::InvalidJWS))?;
+        }
+    }
+
+    let pem = format!(
+        "-----BEGIN PUBLIC KEY-----\n{}\n-----END PUBLIC KEY-----",
+        base64::encode(certs[0].public_key().raw)
+    );
```
Comment (Parth) [inline server/server/src/billing/app_store_service.rs:118]: Are you sure this is required?

File: server/server/src/billing/app_store_service.rs [hunk 23]
```diff
@@ -0,0 +1,131 @@
...
+    let key = DecodingKey::from_ec_pem(pem.as_bytes())?;
```
Comment (Parth) [inline server/server/src/billing/app_store_service.rs:119]: Since this wasn't the main issue, couldn't we use the non-pem one and not turn our `der` into `pem` ourselves?

**Review by Parth — CHANGES_REQUESTED** [review]

File: clients/apple/Shared/Stateful Logic/BillingService.swift (hunk 1 above)
Comment (SmailBarkouch) [reply clients/apple/Shared/Stateful Logic/BillingService.swift:42]: It can be moved, but in the storekit example apple provides, it is called similarly. It shouldn't be a problem to move it though.

File: clients/apple/Shared/Stateful Logic/BillingService.swift (hunk 2 above)
Comment (SmailBarkouch) [reply clients/apple/Shared/Stateful Logic/BillingService.swift:172]: It won't act unexpectedly later. The `maybeMonthlySubscription` is a nullable `Product`. If it is null at the `purchaseProduct` method, then I will simply say "Failed to complete purchase."

File: clients/apple/Shared/Stateful Logic/BillingService.swift (hunk 4 above)
Comment (SmailBarkouch) [reply clients/apple/Shared/Stateful Logic/BillingService.swift:100]: The receipt should [be generated after a purchase](https://developer.apple.com/documentation/appstorereceipts/validating_receipts_with_the_app_store). It's just a matter of waiting.

File: core/libs/shared/src/api.rs (hunk 13 above)
Comment (SmailBarkouch) [reply core/libs/shared/src/api.rs:394]: Original means the first transaction. That is a way you can identify subscriptions.

File: server/server/src/billing/app_store_client.rs (hunk 14 above)
Comment (SmailBarkouch) [reply server/server/src/billing/app_store_client.rs:11]: It cannot, it is a part of [apple's standards](https://developer.apple.com/documentation/appstoreserverapi/generating_tokens_for_api_requests).

File: clients/apple/Shared/Stateful Logic/BillingService.swift (hunk 5 above)
Comment (SmailBarkouch) [reply clients/apple/Shared/Stateful Logic/BillingService.swift:102]: It will be caught by the `catch` on the outside of the loop.

File: clients/apple/Shared/Stateful Logic/BillingService.swift (hunk 6 above)
Comment (SmailBarkouch) [reply clients/apple/Shared/Stateful Logic/BillingService.swift:138]: Doesn't have to but if I did return `nil` I would have to add more lines of code.

File: clients/apple/SwiftLockbookCore/.swiftpm/xcode/xcshareddata/xcschemes/SwiftLockbookCore.xcscheme [hunk 24]
```diff
@@ -1,6 +1,6 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <Scheme
-   LastUpgradeVersion = "1240"
+   LastUpgradeVersion = "1400"
```
[the remark this answers is not in the document]
Comment (SmailBarkouch) [reply clients/apple/SwiftLockbookCore/.swiftpm/xcode/xcshareddata/xcschemes/SwiftLockbookCore.xcscheme:3]: Seems to be a value automatically modified by xcode.

File: clients/apple/SwiftLockbookCore/Sources/SwiftLockbookCore/Models/FileUsage.swift [hunk 25]
```diff
@@ -4,16 +4,32 @@ public struct UsageMetrics: Codable {
...
 }
 
 public struct FileUsage: Codable {
     public var fileId: UUID
     public var sizeBytes: UInt64
+    
+    public init(fileId: UUID, sizeBytes: UInt64) {
+        self.fileId = fileId
+        self.sizeBytes = sizeBytes
+    }
```
[the remark this answers is not in the document]
Comment (SmailBarkouch) [reply clients/apple/SwiftLockbookCore/Sources/SwiftLockbookCore/Models/FileUsage.swift:32]: Seems `Codable` does not let us use the free constructor all `structs` have, so one is manually made. I did not make this addition by the way, it is a relic of the original PR.

File: server/server/src/billing/app_store_service.rs (hunk 23 above)
Comment (SmailBarkouch) [reply server/server/src/billing/app_store_service.rs:119]: I tried to use der and it did not work as expected, likely a little thing I am missing.

File: clients/apple/Shared/Stateful Logic/BillingService.swift (hunk 4 above)
Comment (Parth) [reply clients/apple/Shared/Stateful Logic/BillingService.swift:102]:
```
    func getReceipt() -> String? {
        if let appStoreReceiptURL = Bundle.main.appStoreReceiptURL,
            FileManager.default.fileExists(atPath: appStoreReceiptURL.path) {

            do {
                return try Data(contentsOf: appStoreReceiptURL, options: .alwaysMapped).base64EncodedString()
            }
            catch {
                return nil
            }
        }
        
        return nil
    }
    ```
    
I think waiting for a file to exist in a loop is fine, you should add a bit of a sleep regardless if you're going to do this. Can you link to a sample that does this? Feels strange regardless. But you should have a timeout, if some assumption of yours is wrong this will infinite loop.

1. handle the errors in the catch better
2. sleep and timeout

File: clients/apple/Shared/Stateful Logic/BillingService.swift [hunk 26]
```diff
@@ -0,0 +1,198 @@
...
+                                }
+                            }
+                        } else {
+                            await transaction.finish()
+                        }
+                    case .failure(let error):
+                        DI.errors.handleError(error)
+                    }
+                } catch {
+                    print("Transaction failed verification")
```
Comment (Parth) [inline clients/apple/Shared/Stateful Logic/BillingService.swift:79]: This is never going to make it to the user, we have a way to handle unexpected errors
